File: 2016/Day9/main.lua
---@param input string
---@return number
local function decompressed_length(input)
    local length = 0
    local i = 1

    while i <= #input do
        if input:sub(i, i) == '(' then
            -- Find the end of the marker
            local marker_end = input:find(')', i, true)
            if not marker_end then
                error('Invalid marker at position ' .. i)
            end

            -- Parse the marker
            local marker = input:sub(i + 1, marker_end - 1)
            local x_pos = marker:find('x', 1, true)
            if not x_pos then
                error('Invalid marker format at position ' .. i)
            end

            local chars_to_repeat = tonumber(marker:sub(1, x_pos - 1))
            local repeat_count = tonumber(marker:sub(x_pos + 1))

            -- Move the index past the marker
            i = marker_end + 1

            -- Add the length of the repeated string
            length = length + chars_to_repeat * repeat_count

            -- Move the index past the repeated string
            i = i + chars_to_repeat
        elseif not input:sub(i, i):match('%s') then
            -- Regular character, just count it
            length = length + 1
            i = i + 1
        else
            -- Skip whitespace
            i = i + 1
        end
    end

    return length
end

---@param input string
---@param start number?
---@param finish number?
---@return number
local function decompressed_length_v2(input, start, finish)
    local length = 0
    local i = start or 1
    local end_pos = finish or #input

    while i <= end_pos do
        if input:sub(i, i) == '(' then
            -- Find the end of the marker
            local marker_end = input:find(')', i, true)
            if not marker_end then
                error('Invalid marker at position ' .. i)
            end

            -- Parse the marker
            local marker = input:sub(i + 1, marker_end - 1)
            local x_pos = marker:find('x', 1, true)
            if not x_pos then
                error('Invalid marker format at position ' .. i)
            end

            local chars_to_repeat = tonumber(marker:sub(1, x_pos - 1))
            local repeat_count = tonumber(marker:sub(x_pos + 1))

            -- Move the index past the marker
            i = marker_end + 1

            -- Recursively calculate the length of the repeated string
            local repeated_length = decompressed_length_v2(input, i, i + chars_to_repeat - 1)
            length = length + repeated_length * repeat_count

            -- Move the index past the repeated string
            i = i + chars_to_repeat
        elseif not input:sub(i, i):match('%s') then
            -- Regular character, just count it
            length = length + 1
            i = i + 1
        else
            -- Skip whitespace
            i = i + 1
        end
    end

    return length
end

local file = io.open('input.txt', 'r')
if not file then
    error('Failed to open input file')
end

local input = file:read('*all')
file:close()

print('Part 1: ' .. decompressed_length(input))
print('Part 2: ' .. decompressed_length_v2(input))

File: 2016/Day9/main.py
import sys


def decompossed_length(input: str) -> int:
    length = 0
    i = 0
    n = len(input)

    while i < n:
        if input[i] == '(':
            marker_end_idx = input.index(')', i)
            marker_content = input[i + 1:marker_end_idx]

            x_pos = marker_content.index('x')

            chars_to_repeat = int(marker_content[:x_pos])
            repeat_count = int(marker_content[x_pos + 1:])

            i = marker_end_idx + 1
            length += chars_to_repeat * repeat_count
            i += chars_to_repeat
        elif not input[i].isspace():
            length += 1
            i += 1
        else:
            i += 1

    return length

memo = {}

def decompossed_length_memo(input: str, start_index: int = 0, end_index: int = -1) -> int:
    global memo

    actual_end_index = len(input) if end_index == -1 else end_index
    if start_index >= actual_end_index:
        return 0

    memo_key = (start_index, actual_end_index)
    if memo_key in memo:
        return memo[memo_key]

    current_length = 0
    i = start_index

    while i < actual_end_index:
        if input[i] == '(':
            marker_end_idx = input.find(')', i)
            if marker_end_idx == -1:
                return -1

            marker_content = input[i + 1:marker_end_idx]
            x_pos = marker_content.index('x')

            chars_to_repeat = int(marker_content[:x_pos])
            repeat_count = int(marker_content[x_pos + 1:])

            data_segment_start_abs = marker_end_idx + 1
            data_segment_end_abs = data_segment_start_abs + chars_to_repeat

            repeated_segment_len = decompossed_length_memo(input, data_segment_start_abs, data_segment_end_abs)
            if repeated_segment_len == -1:
                return -1

            current_length += repeated_segment_len * repeat_count
            i = data_segment_end_abs
        elif not input[i].isspace():
            current_length += 1
            i += 1
        else:
            i += 1

    memo[memo_key] = current_length
    return current_length

if __name__ == '__main__':
    input = sys.stdin.read()

    print(f'Part 1: {decompossed_length(input)}')
    memo.clear() # Just to be sure
    print(f'Part 2: {decompossed_length_memo(input)}')
